File: LC_206_reverse_linked_list.rb
# LC 206 Reverse Linked List
# EASY
# Reverse a singly linked list
# https://leetcode.com/problems/reverse-linked-list/
# Input:  1 -> 2 -> 3 -> 4 -> 5 -> NULL
# Output: 5 -> 4 -> 3 -> 2 -> 1 -> NULL

# Definition for singly-linked list.
# class ListNode
#     attr_accessor :val, :next
#     def initialize(val)
#         @val = val
#         @next = nil
#     end
# end

# SOLUTION V1- Iterative- O(n) Time, O(1) Space, where n = list length
# Non-mutative of input
# Takes in head list node, returns new head where subsequent nodes are reversed from original
# @param {ListNode} head
# @return {ListNode}
def reverse_list(head)
  # make copy of head (will be current node)
  currentNode = head

  # make var to hold previous node initialize to null (we will return this)
  previousNode = nil

  # traverse through linked list
  while currentNode != nil
    # make temp copy of next node
    temp = currentNode.next

    # make current node point to previous
    currentNode.next = previousNode

    # set previous to current
    previousNode = currentNode

    # set current to temp
    currentNode = temp
  end

  return previousNode
end

File: LC_206_reverse_linked_list.py
# LC 206 Reverse Linked List
# EASY
# Reverse a singly linked list
# https://leetcode.com/problems/reverse-linked-list/
# Input:  1 -> 2 -> 3 -> 4 -> 5 -> NULL
# Output: 5 -> 4 -> 3 -> 2 -> 1 -> NULL

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

# MUTEATES INPUT HEAD NODE
class Solution:
  def reverseList(self, head: ListNode) -> ListNode:
    # make var to track previous node, initialize to None (will return)
    previous = None

    # traverse linked list (note* head is current node)
    while head is not None:
      # set next var set to next node
      next = head.next

      # make next node point to previous
      head.next = previous

      # set previous node to head (current)
      previous = head

      # set head (current) to next
      head = next

    return previous

